### ISSUE TITLE
Logs information content changes with the format

### ISSUE THREAD

**Issue by pierluigilenoci** [kubernetes/klog#211]
/kind bug

**What steps did you take and what happened:**

Using [`secrets-store-csi-driver`](https://github.com/kubernetes-sigs/secrets-store-csi-driver) (which uses `klog`) I noticed this strange behavior. 

Log with verbosity 1 in text format 
```text
I0205 10:07:14.524904       1 secretproviderclasspodstatus_controller.go:183] "reconcile started" spcps="[redacted namespace]/[redacted secret provider]"
I0205 10:07:14.524921       1 secretproviderclasspodstatus_controller.go:183] "reconcile started" spcps="[redacted namespace]/[redacted secret provider]"
I0205 10:07:14.524938       1 secretproviderclasspodstatus_controller.go:183] "reconcile started" spcps="[redacted namespace]/[redacted secret provider]"
```

Log with verbosity 1 in JSON format
```json
{"ts":1612518712649.7717,"msg":"reconcile started","v":0}
{"ts":1612518712649.785,"msg":"reconcile started","v":0}
{"ts":1612518724148.2576,"msg":"reconcile started","v":0}
{"ts":1612518724148.306,"msg":"reconcile complete","v":0}
{"ts":1612518910965.3506,"msg":"reconcile started","v":0}
{"ts":1612518924913.5608,"msg":"reconcile started","v":0}
```

In JSON `spcps` is missing, `secretproviderclasspodstatus_controller.go:183` is missing

**What did you expect to happen:**

The logs have the same information regardless of the format 

**Anything else you would like to add:**

If you think the problem on `secrets-store-csi-driver` side feel free to comment directly here: https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/442

**Comment by pierluigilenoci** [comment]
@pohly @dims could you please take a look?

**Comment by pohly** [comment]
This may also be an issue with k8s.io/component-base/logs/json and the Zap backend that it uses for JSON. I don't have much personal experience with that myself. I may have some time later this week to investigate, but no promise.

**Comment by pierluigilenoci** [comment]
@pohly I tagged you because I thought you were the right person.
If you are not, let me know who is more correct to tag. ❤️

**Comment by yuzhiquan** [comment]
@pierluigilenoci [pr179](https://github.com/kubernetes/klog/pull/179) dose not resolve this issue?

**Comment by pohly** [comment]
Support for structured logging was added by @munnerz .

I think it is intentional that klog itself leaves it to the structured logger instance to print the source code location. However, because klog wraps such a logger (https://github.com/kubernetes/klog/blob/06bfcd3f996e691aec13671ee2f29323e567fa97/klog.go#L791-L794) the source code information probably would be wrong. `WithCallDepth` was added to logr, so that part can be handled better by klog  now. That still leaves the question whether source code information should be enabled when using the component-base helper code: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/a8e03b4258afa29e733d298233e78ee53391a0f9/cmd/secrets-store-csi-driver/main.go#L80-L82

What is `spcps`? Where does that get added or printed?

**Comment by pierluigilenoci** [comment]
@yuzhiquan yes, it is fixed in klog v2.5.0 but now I'm hit by this problem: https://github.com/kubernetes/klog/issues/212 :( 

@pohly @yuzhiquan to better understand the context: https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/442

**Comment by serathius** [comment]
@pierluigilenoci Thanks for feedback. JSON format is currently in Alpha state and is not planned to be 1to1 replacement to text format. Fields offered by those formats will be different. 

For now we are not planning to add code source information like `secretproviderclasspodstatus_controller.go:183`. As for Alpha we focused on minimal set of required fields and we hope that with migration to structured logging, log messages will become more unique giving an more reliable to find log producer.

If you have any questions regarding Structured Logging or JSON format feel free to direct them to me.
/cc

**Comment by serathius** [comment]
This bug also touched on other issue. Currently JSON is automatically installed in only K8s control plane components, but any external component that depends on k8s.io/component-base.
As an alpha feature this may not be something that developers of external components want to introduce or support. I think we should consider making json installation optional and only force it in control plane components. 

/cc @dims @ehashman @mtaufen
Do you think it is reasonable to allow external K8s component to opt-out of JSON format?

**Comment by mtaufen** [comment]
Technically component-base is about core K8s pieces, and we make clear in the README: 
> There are NO compatibility guarantees for this repository, yet. It is in direct support of Kubernetes, so branches will track Kubernetes and be compatible with that repo.

That said I don't like breaking people, so making the JSON install optional sounds like the right decision to me.

But I do have a question: Is it "installed but not enabled" or "installed and enabled by default" right now? If the latter, it's not really alpha no matter what we say... but if it's the former it's probably ok to change (if we think folks have had sufficient warning of the alpha status).

**Comment by serathius** [comment]
Let's summarize the issue before closing.

Original problem
> In JSON spcps is missing, secretproviderclasspodstatus_controller.go:183 is missing

During the discussion we came to conclusion that installing JSON should be optional for users of component-base.
>That said I don't like breaking people, so making the JSON install optional sounds like the right decision to me.

TODO:
* spcps is missing - Resolved by https://github.com/kubernetes/klog/pull/179
* secretproviderclasspodstatus_controller.go:183 is missing - tracked in https://github.com/kubernetes/kubernetes/issues/102353
* JSON optional installation - Tracked in https://github.com/kubernetes/kubernetes/issues/102632

With all worked tracked, I think we can close this issue. Feel free to open new Issue if there are any new issues related to JSON format and mark it with `wg/structured-logging` label.
/close
